mnefun/_mnefun.py

```diff
@@ -2391,6 +2391,194 @@ def plot_reconstruction(evoked, origin=(0., 0., 0.04)):
+        coeffs_hpi = coeffs[2 + 2*len(linefreqs):]
+        resid_vars[:, epn] = np.var(epdata - np.dot(model, coeffs), 0)
+        # get total hpi amplitudes by combining sine and cosine terms
+        hpi_amps = np.sqrt(coeffs_hpi[0::2, :]**2 + coeffs_hpi[1::2, :]**2)
+        # divide average HPI power by average variance
+        snr_avg_grad[:, epn] = np.divide((hpi_amps**2/2)[:, pick_grad].mean(1),
+                                         resid_vars[pick_grad, epn].mean())
+        snr_avg_mag[:, epn] = np.divide((hpi_amps**2/2)[:, pick_mag].mean(1),
+                                        resid_vars[pick_mag, epn].mean())
+    if hpi_coil == 'median':
+        fun = np.median
+    elif hpi_coil == 'best':
+        fun = np.max
+    elif hpi_coil == 'worst':
+        fun = np.min
+    else:
+        raise ValueError('Invalid HPI coil selection parameter')
+
+    if channels == 'grad':
+        snr = fun(snr_avg_grad, axis=0)
+    elif channels == 'mag':
+        snr = fun(snr_avg_mag, axis=0)
+    return snr
+
+
+def _weigh_epochs(epochs, weights):
+    """ Weigh epochs in mne.Epochs instance (= elementwise multiply by
+    weights vector). """
+    epochs.load_data()
+    weights = np.array(weights)  # will accept either list or numpy array
+    n_epochs = len(epochs)
+    if not len(weights) == n_epochs:
+        raise ValueError('Need as many weights as epochs')
+    # reshape for broadcasting
+    w_ = weights.squeeze()[:, np.newaxis, np.newaxis]
+    # normalize by n_epochs/sum(weights), so that averaging
+    # will yield a correct result
+    epochs._data *= n_epochs * w_ / np.sum(w_)
+
+
+def chpi_weighted_average(raw_snr, raw_epochs=None, reject=False, flat=False,
+                          nharm=None, epoch_start=None, epoch_end=None,
+                          stim_channel=None, mask=None):
+    """ Average epochs weighted by SNR of continuous HPI (cHPI).
+
+    Parameters
+    ----------
+
+    raw_chpi : instance of Raw
+        The raw instance to get the cHPI SNR from. This file must not be
+        MaxFiltered, because MaxFilter removes the cHPI signals.
+    raw_epochs : instance of Raw | None
+        The raw instance to get the data epochs from. If not specified, use
+        epochs from raw_chpi.
+    reject : bool
+        Whether to use the rejection parameters from the data acquisition.
+    flat : bool
+        Whether to use the flatness criteria from the data acquisition.
+    nharm : int
+        Number of line frequency harmonics to use.
+    epoch_start : float
+        Start of epoch relative to the zero time. If not specified, taken from
+        the data acquisition settings.
+    epoch_end : float
+        End of epoch relative to the zero time. If not specified, taken from
+        the data acquisition settings.
+    stim_channel : str
+        The trigger channel to get the events from.
+    mask : int | None
+        The digital mask to apply to the trigger channel values.
```

Comment:
This is not DRY, and also inconsistent with MNE-Python use cases. This break will make many things difficult, including inclusing in the greater `mnefun` processing pattern. A better plan would be to have the user pass their own `events` array